src/services/DexPoolScanner.ts

```diff
@@ -0,0 +1,183 @@
+
+    if (blocksToScan < this.step - 1) {
+      this.logger.debug(`${this.logPrefix} fromBlock: ${fromBlock}, currentBlock: ${currentBlock}, step: ${this.step}`);
+      return {fromBlock, success: false, waitBlockRange: true};
+    }
+
+    const toBlock = this.getToBlock(fromBlock, currentBlock);
+
+    let success = false;
+
+    try {
+      success = await this.apply(fromBlock, toBlock);
+    } catch (e) {
+      console.log(`${this.logPrefix} ${e}`);
+    }
+
+    if (success) await this.updateMarker(toBlock);
+    return {fromBlock, toBlock, success};
+  }
+
+  private async apply(fromBlock: number, toBlock: number): Promise<boolean> {
+    const events = await this.getPoolCreatedEvents(fromBlock, toBlock);
+    if (events.length == 0) return true;
+
+    const addresses = events.map((e) => [e.token1, e.token0]).flat();
+    const sortedEventsByToken = this.sortArrayByTokens(events);
+
+    for (let i = 0; i < sortedEventsByToken.length; i++) {
+      try {
+        const event = sortedEventsByToken[i];
+        const fee = Number(event.fee);
+        const pool = String(event.pool);
+
+        this.logger.info(`${this.logPrefix} Saving token0: ${event.token0}, token1: ${event.token1}. From ${pool}`);
+
+        await this.dexRepository.upsert({
+          chainId: this.chainId,
+          dexProtocol: this.dexProtocolName,
+          token0: event.token0.toLocaleLowerCase(),
+          token1: event.token1.toLocaleLowerCase(),
+          pool,
+          fee,
+        });
+      } catch (e) {
+        this.logger.error(`${this.logPrefix} Could not get a valid symbol for addresses ${addresses} ${e}`);
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private async getPoolCreatedEvents(fromBlock: number, toBlock: number): Promise<PoolCreatedEvent[]> {
+    this.logger.info(`${this.logPrefix} Scanning for Uniswap events between ${fromBlock} to ${toBlock}`);
+    const events = await this.dexProtocol.getPoolCreatedEvents(fromBlock, toBlock);
+    this.logger.info(`${this.logPrefix} Got ${events.length} new Uniswap pools.`);
+    this.logger.debug(`${this.logPrefix} Events: ${JSON.stringify(events)}`);
+    return events;
+  }
+
+  private sortArrayByTokens(events: PoolCreatedEvent[]) {
+    return events.sort((a, b) => {
+      const tokenA0 = a.token0.toLowerCase();
+      const tokenA1 = a.token1.toLowerCase();
+      const tokenB0 = b.token0.toLowerCase();
+      const tokenB1 = b.token1.toLowerCase();
+
+      if (tokenA0 === tokenA1 || tokenB0 === tokenB1) {
+        throw new Error(`${this.logPrefix} tokens are the same, pool: ${a.pool}`);
+      } else {
+        return tokenA0.localeCompare(tokenB0);
+      }
+    });
+  }
+
+  private getToBlock(fromBlock: number, latestBlock: number): number {
+    const stepBlock = fromBlock + this.step - 1;
+
+    // Math.trunc(blocksToScan / this.step) * this.step;
+
+    return latestBlock > stepBlock ? stepBlock : latestBlock;
+  }
```

Comment:
using the internal `step` data member makes it a bit more difficult to unit test this function. Maybe passing it as an argument could enable quick and easy unit testing.

Reply:
Looking for a testing perspective, make sense. But I think it is part of the scanner, and it came from a settings, so I believe it is better in class. For testing is just one more settings, nothing much too